### ISSUE TITLE
Parametrs

### ISSUE THREAD

**Issue by xiaonh** [brichard19/BitCrack#29]
Hello,
How to set maximum parameters for my GTX 1080 ti ?
So it has 3584 cuda blocks so i need to set -b 3584 ?
Maximum number of threads per block:           1024
so -t 1024?
And what is -p value?
If i set this it generate 35000mb file and stopped
Please help to set maximum number value
Thank you

**Comment by virus-cyber** [comment]
-p 256

**Comment by virus-cyber** [comment]
Why do you need this if you can not even configure

**Comment by virus-cyber** [comment]
Why do you need this if you can not even configure

**Comment by brichard19** [comment]
Hi,

I've been meaning to make a tool that finds optimal parameters. I only have a GeForce GT 640 DDR3, so I can only find optimal parameters for that card.

The two important parameters are blocks (`-b`) and threads per block (`-t`).

If you run `cudaInfo.exe` (included in release), it will tell you how many MP's your card has.

The number of blocks  should be a multiple of the number of MP's, since blocks do not move between MP's. You will want the total number of threads (blocks x threads per block) to be at least the number of CUDA cores on your card.


The `-p` parameter is for points per thread: How many points each thread processes simultaneously. This is basically the intensity factor. A high value for `-p` will increase the speed, but it will also increase the running time of the kernel, which could cause crashes if the card is running a display.


Example:

```
C:\BitCrack\cudaInfo.exe
Found 1 devices

ID:          0
Name:        GeForce GT 640
Capability:  3.0
MP:          2
Cores:       384 (192 per MP)
Memory:      1024MB
```

My card has 2 MP's, each with 192 cores. I've found the best performance when I use 32 blocks (`-b 32`), and 256 threads per block (`-t 256`) for 8192 total threads. And I usually use `-p 384`.

I hope this helps.

Thanks

**Comment by xiaonh** [comment]
Thank you for answering.

Found 1 devices

ID:          0
Name:        GeForce GTX 1080 Ti
Capability:  6.1
MP:          28
Cores:       1792 (64 per MP)
Memory:      11264MB


So, what parameters are best for me?

**Comment by brichard19** [comment]
Hi,

I don't have a 1080 Ti, so I can't say what the best parameters are. But it will be at least 28 blocks 64 threads per block.

So try combinations of 28, 56, 84, 112 for `-b` and 64, 128, 192, 256, 320... for `-t` and see if you can find which gives you the most speed e.g. maybe 112 blocks and 256 threads per block.

Once you find that, increase the `-p` value.

Hope that helps.

Thanks

**Comment by bill32767** [comment]
C:\Users\copo\Desktop\BitCrack-master>bitcrack64
[OPTIONS] [TARGETS]
Where TARGETS is one or more addresses

Integer arguments can be in decimal (e.g. 123) or hex (e.g. 0x7B or 7Bh)

-c, --compressed        Compressed points
-u, --uncompressed      Uncompressed points
-d, --device            The device to use
-b, --blocks            Number of blocks
-t, --threads           Threads per block
-p, --per-thread        Keys per thread
-s, --start             Staring key, in hex
-r, --range             Number of keys to search
-i, --in                Specify file containing addresses, one per line
-o, --out               Specify file where results are written

C:\Users\copo\Desktop\BitCrack-master>

PLEASE EXPLAIN THIS: *Integer arguments can be in decimal (e.g. 123) or hex
(e.g. 0x7B or 7Bh)*
AND WHY THE *PK* THE TOOL SEARCHES IS NUMERICAL e.g:
(1LbGyuYV6BgBv6LLf7jdHUfu98YmTZ4LqE
0000000000000000000000000000000000000000000000000000000000005000) AND NOT
ALPHA-NUMERICAL LIKE THE EXAMPLE YOU ABOVE IN bitcrack64 (*e.g. 0x7B.....
or 7Bh.......*)

Thanks.

**Comment by natedawg469** [comment]
@brichard19 How long does it take for your bitcrack program to actually start working? Mine was stuck for over 24 hours
![bitcrack64 taking super long](https://user-images.githubusercontent.com/32862997/45071215-30d19700-b0a3-11e8-97c6-994e810dbc4a.png)

**Comment by natedawg469** [comment]
Also what would happen if I overclock GPU in EVGA Precision XOC?

**Comment by natedawg469** [comment]
Also what does it actually software wise and internally to the GPU when it is "verifying points on device"?

**Comment by brichard19** [comment]
@natedawg469 You built in Debug mode. Build in Release mode to avoid this. In Debug mode it performs a check to make sure the GPU initialized correctly.

**Comment by adminnu** [comment]
I'm using -p 376 but it's 256 and it shows the same result, just keeps the speed a little more stable. The result is at 1080 ti ~ 440 mk / s (+ - 20 mk / s)

**Comment by brichard19** [comment]
@adminnu `-p` increases performance asymptotically: it approaches a maximum.

440 mkeys/sec sounds impressive though. It's much more than what my GPU can do.

**Comment by natedawg469** [comment]
Holy shit! I have 6 EVGA GTX 1080 FTWs, so basically you guys are telling me that if I can get this shit working correctly, I basically have a mini super computer. I onow 3584 cores is crazy. Im also running 32gb GDDRX5 ram on the system.

**Comment by xiaonh** [comment]
so, what are the best settings for 1080 ti? my maximum that i get is 220mk/s.
How you get 440?

**Comment by natedawg469** [comment]
@adminnu what is your -b figure you are running?

**Comment by natedawg469** [comment]
Thanks @brichard19 
![bitcrack fixed](https://user-images.githubusercontent.com/32862997/45107699-6d83a980-b108-11e8-8862-9de302b69d9c.png)
for your help! Got in working and it started in less than 20 seconds

**Comment by natedawg469** [comment]
Interesting with messing with the parameters. I increased the threads and it lowered the keys/s
![bitcrack slower speed](https://user-images.githubusercontent.com/32862997/45109056-65c60400-b10c-11e8-9cd4-0416e518ef4e.png)

**Comment by Relok1995** [comment]
Guys, who ever uses gtx 950? Or just 9 generations? What parameters do you use?

**Comment by natedawg469** [comment]
Why when I increase threads and -p it makes like for example 8,800,461 mkeys/s and so I run that from hex of address 55 (which 30 quadrillion something) all the way to 2^56 and it doesn't find 17aPYR... (address 56). It literally does not find the private key to the address. Does not make sense to me.

**Comment by Relok1995** [comment]
I assume that your gpu can not calculate more by the given parameters, hence BitCrack gives you false values ​​of mk / s

**Comment by natedawg469** [comment]
@Relok1995 Here is what I got: 
![keys](https://user-images.githubusercontent.com/32862997/45140494-f9391c80-b180-11e8-809d-bd042d3e93e7.png)

**Comment by adminnu** [comment]
> @adminnu, какова ваша цифра -b, которую вы используете?

-b I do not use, since-b and -t are the same as -p. That is, -p is constructed from -b and -t, if you specify the last -p, it is ignored.
I do not know, but I have five msi 1080 ti gaming x, the same value in the 440 area, it's a little less, it's a bit more. And this value is from -p 256, I just noticed that at 376 and 312 slightly more securely keeps the speed level. I also did not test whether it finds the address at 376, tested it at 256, it's bad if yesterday, changing this value, it already does not find it, read the comment above.

What can I say, why so much, maybe you have 1080 ti of another manufacturer, for example I have one msi 1080 ti just gaming without x, it gives slightly lower values, although I did not check it, because now the second power supply is burned out, I'll go today it is repaired. Still maybe you do not have overclocking cards, I used an underworld, that is, I have all the cards held at 975 mv with a frequency of about 1960 - 2100 mhz and I have a lot of memory for each card.

**Comment by natedawg469** [comment]
@adminnu  Quote: "Not, on 376, too, he found, added to the starting number a trillion in the decimal sister, put this address to search and he found, as expected, in 2-3 seconds."

Who are you talking to and what are you talking about?

**Comment by adminnu** [comment]
@natedawg469 you yourself asked what parameter I used?

**Comment by brichard19** [comment]
@natedawg469 that 8 million mkeys/s is  definitely a bug.

**Comment by bill32767** [comment]
I thought as much. Can it be fixed? At least CudaInfo should say the best
least/most working parameters based on the GPU properties. Example: GPU's
with more than one -d will work faster and process more keys at a maximum
number of -b -t -p based on there respective mathematical calculations.

**Comment by adminnu** [comment]
@bill32767 I'll upset you, bitcoin crackers. If you pay attention to the map at the time of this hyper speed, you will see that its load is 0%. This happens when the -t parameter increases, more than 512.

**Comment by bill32767** [comment]
Okay. Noted.

**Comment by natedawg469** [comment]
Who found puzzle key 56?

**Comment by bill32767** [comment]
I thought you did based on your previous comment "Holy fuck! I broke
bitcoins".....

**Comment by natedawg469** [comment]
No um some guy named arulbero in the bitcoin talk page is using some sort decimal/binary programming program he built and so far I believe he is the one who emptied 55 and 56.

**Comment by bill32767** [comment]
How did you know about this???

**Comment by bill32767** [comment]
I'm seeing this for the first time. I have no idea.....

**Comment by natedawg469** [comment]
Im trying to beat time. I realize what I may be trying to do is impossible, but waiting 30 days to scan 1/4 of the of quadrillions range of keys were in takes forever.

**Comment by brichard19** [comment]
What happens when t >= 512?

I wonder if we are running into limits on the number of threads.

**Comment by natedawg469** [comment]
Hey thanks @brichard19 for turning this into a Parameters discussion. Like I said before in recent posts I do not know how to code but I can tell one thing is that 6 state of the art GTX 1080 ftw graphics with tremendous computing capabilities and so I believe too that there is something limiting the threads, Also I am thinking that maybe the GPUs are running through false points too when you raise the threads -t and -p higher.

**Comment by brichard19** [comment]
@natedawg469 Can you provide an example input that shows the bug (with target address, starting key, parameters etc)? I am not seeing any problem running with `-t 512`. But I only have a GT 640 to test on.

**Comment by natedawg469** [comment]
![not finding address](https://user-images.githubusercontent.com/32862997/45565755-956abf80-b822-11e8-9408-023c1f2f0a08.png)

@brichard19  here is this photo I set the target address as 1KoNcTGrYsEhpMEsLJgTTCQwnXiSRLx5BH which I think is only about 2 trillion keys away from the starting point. So with these parameters in place, bitcrack does not find the key. It goes past it.

**Comment by brichard19** [comment]
That's very odd. I find it almost instantly on my system:

I am unable to test with -b 320 because I only have 1GB memory. Can someone else verify this issue?

'''
D:\BitCrack>BitCrack64.exe 1KoNcTGrYsEhpMEsLJgTTCQwnXiSRLx5BH -b 32 -t 640  -p 200 -s 10009184e72a000 -r 72057594037927927936
[2018-09-14.18:51:06] [Info] Compression: compressed
[2018-09-14.18:51:06] [Info] Starting at: 000000000000000000000000000000000000000000000000010009184E72A000
[2018-09-14.18:51:06] [Info] Initializing GeForce GT 640
[2018-09-14.18:51:07] [Info] Generating 4,096,000 starting points (156.3MB)
[2018-09-14.18:51:13] [Info] 10.0%
[2018-09-14.18:51:14] [Info] 20.0%
[2018-09-14.18:51:16] [Info] 30.0%
[2018-09-14.18:51:17] [Info] 40.0%
[2018-09-14.18:51:18] [Info] 50.0%
[2018-09-14.18:51:19] [Info] 60.0%
[2018-09-14.18:51:20] [Info] 70.0%
[2018-09-14.18:51:21] [Info] 80.0%
[2018-09-14.18:51:23] [Info] 90.0%
[2018-09-14.18:51:24] [Info] 100.0%
[2018-09-14.18:51:24] [Info] Done
[2018-09-14.18:51:24] [Info] Address:     1KoNcTGrYsEhpMEsLJgTTCQwnXiSRLx5BH
                             Private key: 000000000000000000000000000000000000000000000000010009184E81E460
                             Compressed:  yes
                             Public key:
                             03C68FF88E22D4E33AC603F6E9982D7325A03913F75896F253562D73C6111A29E1


D:\Users\BEN\Documents\GitHub\BitCrack\x64\Release>
'''

Does it work properly if you use smaller parameters?

**Comment by natedawg469** [comment]
@brichard19 yes it works with smaller parameters for -t threads and -p points but it will take very long to find the key. One thing  I have not tried though is lowering -b blocks.

**Comment by natedawg469** [comment]
yeah thats extremely odd, I just did exactly what you did and it still does not find it.

**Comment by bill32767** [comment]
You have six -d and you are running on all, use some -d (like: -d 0 -d 2 -d
3 -d 5) only then you'll see the key.

**Comment by bill32767** [comment]
Also decrease -b -t and increase -p

**Comment by natedawg469** [comment]
yeah you are right @bill32767  it found it instantly
![found it instantly](https://user-images.githubusercontent.com/32862997/45580806-5872ed00-b863-11e8-9f93-6410c7c73d9b.png)

**Comment by brichard19** [comment]
Hi,

I looked through the code and found some places where we were not checking for errors.

Try running the latest release. I wouldn't be surprised if `-b 320 -t 640` gives a runtime error.

**Comment by natedawg469** [comment]
@brichard19 Oh yeah it does no doubt. if i set -b 320 -t 640 -p 250 + , it say not enough memory and exit program. Which I guess makes sense but I am not sure.

**Comment by natedawg469** [comment]
@brichard19 illegal memory access

**Comment by adminnu** [comment]
@brichard19 now he gives up with -t 512 error:
> Error: out of memory Exiting.
What kind of memory is missing, I have 4GB of RAM and a very large swap or is it a memory card?

@nightwalker89 
At 57 range, the speed became lower, somewhere around 370-400 mk/s. 

@brichard19 
What is very strange, is there a dependence of the speed and width of the keys?

**Comment by hamnaz** [comment]
![bitcrack](https://user-images.githubusercontent.com/42878992/45923671-aa6fde80-bf05-11e8-92f6-294ae8372076.png)

**Comment by hamnaz** [comment]
above pic, tesla p-100
without parameters -b -t -p loaded 20m addresses, 142mkey/s
with parameters aprox 200mkey/s
singal address with parameters 414 mkey/s

**Comment by brichard19** [comment]
@adminu No, the length of the key should not affect performance.

**Comment by hamnaz** [comment]
new parameters tested
![bitc](https://user-images.githubusercontent.com/42878992/45946571-b1cce000-c00a-11e8-9b3c-2e6e3d4a357a.png)

**Comment by hamnaz** [comment]
i want test random keys, if you build with random key option too, 
and option brainwallet file input , key ( wordlist(sha256) )
above parameters 225 mkeys/s on both compression
let see what happen on random key generation and/or load wordlist file as brainwallet

**Comment by g00dwin555** [comment]
I change the parameters -b -t -p, I can not get values greater than 350Mkey/s.
What could be the reason? Or is this normal?

-b 280 -t 256 -p 1024

GeForce GTX 1080ti 8671/11264MB | 1 target 352.32 MKey/s
